### PR TITLE
Ensure lvalue struct receivers are not copied by value by implicit Range indexer rewrite

### LINKED ISSUE #65586: Side effects of implicit Range indexer on a struct type are not propagated to the receiver
Compile and run:
```
using System;

class Program
{
    static S s = new S();
    
    static void Main()
    {
        _ = s[1..2];
        Console.WriteLine(s.count); 
    }
}

struct S
{
    public int count;
    
    public int Slice(int start, int length)
    {
        count++;
        return 0;
    }
    
    public int Length => 10;
}
```

Observed output: 0

Expected output: 1

Change `S` to a class and observe expected output produced. 

While fixing local rewrite for this issue, we should also make sure generic scenarios similar to https://github.com/dotnet/roslyn/issues/63221 are handled properly as well.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #65586.
Related to #63221.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenCallTests.cs [hunk 1]
```diff
@@ -28674,6 +28674,2028 @@ .maxstack  4
...
+  IL_000e:  constrained. ""T""
+  IL_0014:  callvirt   ""int IMoveable.Length.get""
+  IL_0019:  ldloc.1
+  IL_001a:  sub
+  IL_001b:  constrained. ""T""
+  IL_0021:  callvirt   ""IMoveable IMoveable.Slice(int, int)""
+  IL_0026:  pop
+  IL_0027:  ret
+}
+");
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenCallTests.cs:28912]: Should we include `VerifyIL("Shift2<T>", ...);` as well here, and in `GenericTypeParameterAsReceiver_ImpicitRangeIndexer_RangeExpression_Struct_Ref`?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenCallTests.cs:28912]: > Should we include ```VerifyIL("Shift2<T>", ...);``` as well here, and in ...

That method is the same as in the "Class" flavor of the test and it's IL is verified there.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenCallTests.cs:28912]: Here we only need to verify its runtime behavior with an input which is a struct

**Review by cston — APPROVED** [review]

**Pull request merged**